### PR TITLE
add prev/next node postprocessor

### PR DESCRIPTION
**Pull request by jerryjliu**
- add "manual" version: user has to specify whether to go forwards/backwards
- add "auto" version: we can infer whether to keep processing nodes ahead or before 

- changed nodeparser to ABC from Protocol, it was breaking pydantic validation

### CODE REVIEW

File: gpt_index/indices/postprocessor/node.py [hunk 1]
```diff
@@ -82,3 +90,229 @@ def postprocess_nodes(
                 new_nodes.append(node)
 
         return new_nodes
+
+
+class BasePrevNextNodePostProcessor(BaseNodePostprocessor):
```
Comment (Disiok) [inline gpt_index/indices/postprocessor/node.py:95]: nit: I prefer not to use inheritance for code re-use. I think we should keep the hierarchy as flat as possible unless you have a good reason.

File: gpt_index/indices/postprocessor/node.py [hunk 2]
```diff
@@ -82,3 +90,229 @@ def postprocess_nodes(
...
+        docstore (DocumentStore): The document store.
+        num_nodes (int): The number of nodes to return (default: 1)
+        mode (str): The mode of the post-processor. Can be "forward" or "backward".
+
+    """
+
+    docstore: DocumentStore
+    num_nodes: int = Field(default=1)
+
+    def _get_forward_nodes(self, node: Node) -> Dict[str, Node]:
```
Comment (Disiok) [inline gpt_index/indices/postprocessor/node.py:113]: related to above: these don't necessarily need to be instance methods attached to a superclass.

Comment (Disiok) [reply gpt_index/indices/postprocessor/node.py:113]: They can be standalone methods in the module, with docstore and num_modes arguments passed in.

File: gpt_index/indices/query/base.py [hunk 3]
```diff
@@ -248,7 +248,10 @@ def retrieve(self, query_bundle: QueryBundle) -> List[NodeWithScore]:
         similarity_tracker = SimilarityTracker()
         nodes = self._retrieve(query_bundle, similarity_tracker=similarity_tracker)
 
-        postprocess_info = {"similarity_tracker": similarity_tracker}
+        postprocess_info = {
+            "similarity_tracker": similarity_tracker,
+            "query_bundle": query_bundle,
+        }
```
Comment (Disiok) [inline gpt_index/indices/query/base.py:254]: I would advise against passing arguments via unstructured dict.

This makes the interface leaky, and hard to reason.

**Review by Disiok — APPROVED** [review]

File: gpt_index/indices/query/base.py (hunk 3 above)
Comment (jerryjliu) [reply gpt_index/indices/query/base.py:254]: fair. would you recommend just passing the query string as an argument then?

File: gpt_index/indices/postprocessor/node.py (hunk 1 above)
Comment (jerryjliu) [reply gpt_index/indices/postprocessor/node.py:95]: yeah, makes sense

**Pull request merged**